ethcore/light/src/transaction_queue.rs

```diff
@@ -0,0 +1,421 @@
+enum CurrentNonce {
+	// Assumed current nonce.
+	Assumed(U256),
+	// Known current nonce.
+	Known(U256),
+}
+
+impl CurrentNonce {
+	// whether this nonce is assumed
+	fn is_assumed(&self) -> bool {
+		match *self {
+			CurrentNonce::Assumed(_) => true,
+			CurrentNonce::Known(_) => false,
+		}
+	}
+
+	// whether this nonce is known for certain from an external source.
+	fn is_known(&self) -> bool {
+		!self.is_assumed()
+	}
+
+	// the current nonce's value.
+	fn value(&self) -> &U256 {
+		match *self {
+			CurrentNonce::Assumed(ref val) => val,
+			CurrentNonce::Known(ref val) => val,
+		}
+	}
+}
+
+// transactions associated with a specific account.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct AccountTransactions {
+	// believed current nonce (gotten from initial given TX or `cull` calls).
+	cur_nonce: CurrentNonce,
+	current: Vec<PendingTransaction>, // ordered "current" transactions (cur_nonce onwards)
+	future: BTreeMap<U256, PendingTransaction>, // "future" transactions.
+}
+
+impl AccountTransactions {
+	fn is_empty(&self) -> bool {
+		self.current.is_empty() && self.future.is_empty()
+	}
+
+	fn next_nonce(&self) -> U256 {
+		self.current.last().map(|last| last.nonce + 1.into())
+			.unwrap_or_else(|| *self.cur_nonce.value())
+	}
+
+	// attempt to move transactions from the future queue into the current queue.
+	fn adjust_future(&mut self) {
+		let mut next_nonce = self.next_nonce();
+
+		loop {
+			match self.future.remove(&next_nonce) {
+				Some(tx) => self.current.push(tx),
+				None => break,
+			}
+
+			next_nonce = next_nonce + 1.into();
+		}
+	}
+}
+
+/// Light transaction queue. See module docs for more details.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct TransactionQueue {
+	by_account: HashMap<Address, AccountTransactions>,
+	by_hash: H256FastMap<PendingTransaction>,
+}
+
+impl TransactionQueue {
+	/// Import a pending transaction to be queued.
+	pub fn import(&mut self, tx: PendingTransaction) -> Result<TransactionImportResult, TransactionError>  {
+		let sender = tx.sender();
+		let hash = tx.hash();
+		let nonce = tx.nonce;
+
+	    let res = match self.by_account.entry(sender) {
+			Entry::Vacant(entry) => {
+				entry.insert(AccountTransactions {
+					cur_nonce: CurrentNonce::Assumed(nonce),
+					current: vec![tx.clone()],
+					future: BTreeMap::new(),
+				});
+
+				TransactionImportResult::Current
+			}
+			Entry::Occupied(mut entry) => {
+				let acct_txs = entry.get_mut();
+				if &nonce < acct_txs.cur_nonce.value() {
+					// don't accept txs from before known current nonce.
+					if acct_txs.cur_nonce.is_known() {
+						return Err(TransactionError::Old)
+					}
+
+					// lower our assumption until corrected later.
+					acct_txs.cur_nonce = CurrentNonce::Assumed(nonce);
+				}
+
+				match acct_txs.current.binary_search_by(|x| x.nonce.cmp(&nonce)) {
+					Ok(idx) => {
+						trace!(target: "txqueue", "Replacing existing transaction from {} with nonce {}",
```

Comment:
If the user wants to try with a lower gas price then they're welcome to, the least we can do is allow our implementation to propagate it to the network. What the peers do with the transaction is beyond our control. The fact that it would probably not work should be left to documentation instead of behavior.